contrib/win32/win32compat/win32_usertoken_utils.c

```diff
@@ -382,4 +392,262 @@ int load_user_profile(HANDLE user_token, char* user) {
+	 PUNICODE_STRING pAccountName,
+	 PSID pSid,
+	 PLSA_SID_NAME_MAPPING_OPERATION_ERROR pResult)
+{
+	LSA_SID_NAME_MAPPING_OPERATION_INPUT   OpInput = { 0 };
+	PLSA_SID_NAME_MAPPING_OPERATION_OUTPUT pOpOutput = NULL;
+	LSA_SID_NAME_MAPPING_OPERATION_ERROR Result =
+		LsaSidNameMappingOperation_NonMappingError;
+	NTSTATUS status = STATUS_SUCCESS;
+	int ret = 0;
+
+	OpInput.AddInput.DomainName = *pDomainName;
+	if (pAccountName)
+		OpInput.AddInput.AccountName = *pAccountName;
+	OpInput.AddInput.Sid = pSid;
+
+	status = LsaManageSidNameMapping(LsaSidNameMappingOperation_Add,
+		&OpInput,
+		&pOpOutput);
+	if (status != STATUS_SUCCESS) {
+		ret = -1;
+		if (pOpOutput) {
+			Result = pOpOutput->AddOutput.ErrorCode;
+			if (Result == LsaSidNameMappingOperation_NameCollision || Result == LsaSidNameMappingOperation_SidCollision)
+				ret = 0; /* OK as it failed due to collision */
+			else
+				error("LsaManageSidNameMapping failed with : %s \n", LSAMappingErrorDetails[Result]);
+
+			LsaFreeMemory(pOpOutput);
+		}
+		else
+			error("LsaManageSidNameMapping failed with ntstatus: %d \n", status);
+	}
+
+	return ret;
+}
+
+
+int RemoveVirtualAccountLSAMapping(PWSTR virtualAccountName)
+{
+	int ret = 0;
+
+	LSA_SID_NAME_MAPPING_OPERATION_INPUT         OpInput = { 0 };
+	PLSA_SID_NAME_MAPPING_OPERATION_OUTPUT       pOpOutput = NULL;
+	PLSA_SID_NAME_MAPPING_OPERATION_REMOVE_INPUT pRemoveSidMapping = &OpInput.RemoveInput;
+
+	RtlInitUnicodeString(&pRemoveSidMapping->DomainName, VIRTUALUSER_DOMAIN);
+	if (virtualAccountName)
+		RtlInitUnicodeString(&pRemoveSidMapping->AccountName, virtualAccountName);
+
+	NTSTATUS status = LsaManageSidNameMapping(LsaSidNameMappingOperation_Remove,
+		&OpInput,
+		&pOpOutput);
+	if (status != STATUS_SUCCESS)
+		ret = -1;
+
+	if (pOpOutput)
+		LsaFreeMemory(pOpOutput);
+
+	return ret;
+}
+
+HANDLE generate_sshd_virtual_token()
+{
+	NTSTATUS status = STATUS_SUCCESS;
+	SID_IDENTIFIER_AUTHORITY NtAuthority = SECURITY_NT_AUTHORITY;
+	LSA_SID_NAME_MAPPING_OPERATION_ERROR result;
+	UNICODE_STRING Domain;
+	UNICODE_STRING Account;
+
+	PWSTR pVirtualAccountUserName = NULL;
+	PSID pSidDomain = NULL;
+	PSID pSidUser = NULL;
+	HANDLE hToken = 0, hRestrictedToken = 0;;
+
+	pVirtualAccountUserName = (PWSTR)malloc(sizeof(WCHAR)*(VIRTUALACCOUNT_NAME_LENGTH_MAX + 1));
+	if (NULL == pVirtualAccountUserName)
+		goto Cleanup;
+
+	ZeroMemory(pVirtualAccountUserName, sizeof(WCHAR)*(VIRTUALACCOUNT_NAME_LENGTH_MAX + 1));
+
+	StringCchPrintfW(
+		pVirtualAccountUserName,
+		(VIRTUALACCOUNT_NAME_LENGTH_MAX + 1),
+		L"%s_%d",
+		L"sshd",
+		GetCurrentProcessId());
+
+	RtlInitUnicodeString(&Domain, VIRTUALUSER_DOMAIN);
+	RtlInitUnicodeString(&Account, pVirtualAccountUserName);
+
+	/* 
+	 * setup domain SID - 
+	 * Windows team recommends using S-1-5-111 named "Virtual Users"
+	 * for all virtual accounts 
+	 */
+	status = AllocateAndInitializeSid(&NtAuthority,
+		1,
+		111,
+		0,
+		0,
+		0,
+		0,
+		0,
+		0,
+		0,
+		&pSidDomain);
+	if (status != TRUE)
```

Comment:
As per [MSDN page ](https://msdn.microsoft.com/en-us/library/windows/desktop/aa375213(v=vs.85).aspx), AllocateAndInitializeSid() returns non-zero if succeeds  so the check here should be if(status == FALSE)?

Reply:
Will fix it.